docs/proposals/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,321 @@
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: guestbook
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-deployer
+  - namespace: guestbook-dev
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-dev-deployer
+  - namespace: guestbook-stage
+    server: https://kubernetes.default.svc
+    serviceAccountName: guestbook-stage-deployer
+  - namespace: *
+    server: https://kubernetes.default.svc
+    serviceAccountName: generic-deployer
+```
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+#### Use case 1:
+
+As a user, I would like to use kubernetes security constructs to restrict user access for application sync
+So that, I can provided granular permissions based on the principal of least privelage required for syncing an application.
+
+#### Use case 2:
+
+As a user, I would like to configure a common service account for all applications associated to an AppProject
+So that, I can use a generic convention of naming service accounts and avoid associating the service account per application.
+
+### Implementation Details/Notes/Constraints
+
+#### Component : GitOps Engine
+
+Fix GitOps Engine code to honor Impersonate configuration set in the Application sync context for all kubectl commands that are being executed.
+
+#### Component: ArgoCD API
+
+- Support new property `serviceAccountName` in `Destination` section of `AppProject`
+- Documentation of the new fields
+
+#### Component: ArgoCD Application Controller
+
+- Provide a configuration in argocd-cm  which can be modified to enable the Impersonation feature. Set applicationcontroller.enable.impersonation: true in the Argo CD ConfigMap.
+- Provide an option to override the Impersonation feature using environment variables.
+- Pass `--enable-impersonation` to the Application controller args.
+- Set `ARGOCD_APPLICATION_CONTROLLER_ENABLE_IMPERSONATION=true` in the Application controller environment variables.
+- Fix Application Controller sync.go to set the Impersonate configuration from the Application CR to the SyncContext Object (rawConfig and restConfig field, need to understand which config is used for the actual sync and if both configs need to be impersonated.)
+#### Component: ArgoCD UI
+
+- Provide option to provide `serviceAccountName` for a `Destination` section in an `AppProject` created/updated via the web console.
+- Update the User Guide documentation on how to use these newly added fields from the web console.
+#### Component: ArgoCD CLI
+
+- Provide option to provide serviceAccountName for an Application created/updated via the console.
+- Provide option to provide defaultServiceAccount for an AppProject created/updated via the web console.
+- Update the User Guide and other documentation where the CLI option usages are explained.
+
+
+### Detailed examples
```

Comment:
I think it would be helpful to add a "how-to" example. Basically a step-by-step guide for a user to take advantage of this hypothetical feature. e.g.

1. Create an AppProject for team A
2. Add a destination for team A's cluster, https://team-a.kubernetes.svc
3. Add a serviceAccountName `team-a-service-account` to the destination
4. Create a cluster config that looks like X and do Y to allow that cluster config to impersonate `team-a-service-account` on the destination cluster
5. etc...

I'm not super familiar with the mechanics of impersonation, so the more detailed drill-down will help me imagine exactly how this feature will work.

Reply:
Also, how would the service account itself be managed? I suspect there'll need to be an impersonation opt-out (something like an annotation on the relevant manifest itself perhaps?), otherwise the service account would not be able to be managed in a Gitops-manner.